azkaban-common/src/main/java/azkaban/trigger/builtin/BasicTimeChecker.java

```diff
@@ -179,11 +230,31 @@ private long calculateNextCheckTime() {
           break;
         }
       } else {
+        // period is deprecated Schedule API parameter, we use cronSchedule API instead.
         date = date.plus(this.period);
       }
       count += 1;
     }
-    return date.getMillis();
+    // it's possible the above while loop exits on break condition instead of that "date refreshed after Now"
+    // this happens when trigger expires before NOW, in such case we remove all missed checkTimes as they're meaningless
+    if (!date.isAfterNow()) {
+      nextCheckTimeObj.missedCheckTimeBeforeNow.clear();
+    }
+    nextCheckTimeObj.nextValidCheckTimeFromNow = date.getMillis();
+    LOG.debug("Done calculation, current nextCheckTime {} and missScheduleList {}",
+        nextCheckTimeObj.nextValidCheckTimeFromNow,
+        nextCheckTimeObj.missedCheckTimeBeforeNow);
```

Comment:
It's hard to get schedule id from top levels. As I said TriggerManager's design is not in good form.
A trigger(schedule) would have trigger_id, but it's in trigger level. A Trigger has triggerCondition, and in triggerCondition it has a map, in the map then is a checker. Checker itself has an ID but it's useless as every condition would have a checker with id "BasicTimeChecker_1". In both checker's json/DB or even condition level we don't know schedule id.
So refactoring scheduleID itself is a dirty work (It will mess up even more structure, if we have to do that I would rather do it in a separate PR so the structure change would be more clear)
For now I think cronExpression can be put in as a temporary identifier. 
I've tested in mlearn2, if I know the nextTime then it does not hard to identity the testing schedule because TriggerManager is managed by a single thread. In this way, no other nextCheckTime calculation would interfere inside current tracking.